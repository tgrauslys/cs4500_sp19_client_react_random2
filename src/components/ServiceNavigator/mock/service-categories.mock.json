[
  {
    "id": "123", "title": "Home Improvements",
    "services": [
      {"id": "1", "title": "Appliance Installation"},
      {"id": "2", "title": "Cabinetry"},
      {"id": "3", "title": "Door Installation"},
      {"id": "4", "title": "Concrete Removal"},
      {"id": "5", "title": "Dishwasher Installation"},
      {"id": "6", "title": "Finish Carpentry"},
      {"id": "7", "title": "Floor Polishing"},
      {"id": "8", "title": "Furniture Assembly"},
      {"id": "9", "title": "Generator Installation"},
      {"id": "10", "title": "Gardening"},
      {"id": "11", "title": "Home Organizing"},
      {"id": "12", "title": "Electrical and Wiring Repair"}
    ],
    "serviceProviders": [
      {
        "title": "Christina's Custom Cleaning, CCC",
        "rating": 4.4,
        "ratingScores": [
          93,
          24,
          12,
          10,
          4
        ],
        "yearsInBusiness": "7",
        "hires": 17,
        "latestReview": "It's so wonderful to have the burden of cleaning - especially with an injury - removed. Christina is excellent with communication and I've been very pleased with the quality of work.",
        "price": "$75",
        "introduction": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum",
        "backgroundChecked": true,
        "employees": 2,
        "paymentMethods": [
          "Cash", "Venmo"
        ],
        "faqs": [
          {
            "question": "What should the customer know about your pricing (e.g., discounts, fees)?",
            "answer": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum"
          },
          {
            "question": "What education and/or training do you have that relates to your work?",
            "answer": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum"
          }
        ],
        "reviews": [
          {
            "id": 1,
            "who": "Kristine M.",
            "stars": 5,
            "review":"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            "date": "Apr 13, 2017",
            "reply": "thank you very much!! see you next weekend!!"
          },
          {
            "id": 2,
            "who": "Eric G.",
            "stars": 5,
            "review":"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
            "date": "Apr 19, 2017",
            "reply": "Thank you so much Eric! We loved cleaning your home! "
          }
        ]
      },
      {
        "title": "Perfect Shine Cleaner",
        "rating": 5,
        "reviews": 1,
        "yearsInBusiness": "1",
        "hires": 3,
        "latestReview": "Ana came is and did an amazing job cleaning before a family party. She was available on short notice and exceeded my expectations. Thanks Ana!",
        "price": "$90"
      },
      {
        "title": "Mendes cleaning services",
        "rating": 4.8,
        "reviews": 35,
        "yearsInBusiness": "5",
        "hires": 23,
        "latestReview": "Wonderful job! House was quite a mess and they did a great job! Scheduling them for regular cleanings! High recommend!!!",
        "price": "$90"
      }
    ],
    "questions": [
      {
        "id": "1",
        "question": "Number of bedrooms",
        "questionType": "MULTIPLE-CHOICE",
        "choices": [
          {"id": "1", "title": "1 bedroom"},
          {"id": "2", "title": "2 bedroom"},
          {"id": "3", "title": "3 bedroom"},
          {"id": "4", "title": "4 bedroom"},
          {"id": "5", "title": "5 bedroom"},
          {"id": "6", "title": "6 bedroom"}
        ]},
      {"id": "2", "question": "Number of bathrooms", "questionType": "MULTIPLE-CHOICE",
        "choices": [
          {"id": "1", "title": "1 bathroom"},
          {"id": "2", "title": "2 bathroom"},
          {"id": "3", "title": "3 bathroom"},
          {"id": "4", "title": "4 bathroom"},
          {"id": "5", "title": "5 bathroom"},
          {"id": "6", "title": "6 bathroom"}
        ]},
      {"id": "3", "question": "Cleaning type", "questionType": "MULTIPLE-CHOICE",
        "choices": [
          {"id": "1", "title": "Standard cleaning"},
          {"id": "2", "title": "Deep cleaning"},
          {"id": "3", "title": "Move out cleaning"}
        ]},
      {"id": "4", "question": "Extra services", "questionType": "MULTIPLE-ANSWER",
        "choices": [
          {"id": "1", "title": "Fridge cleaning"},
          {"id": "2", "title": "Oven cleaning"},
          {"id": "3", "title": "Window cleaning"},
          {"id": "4", "title": "Laundry"}
        ]},
      {"id": "5", "question": "Frequency", "questionType": "MULTIPLE-CHOICE",
        "choices": [
          {"id": "1", "title": "Just once"},
          {"id": "2", "title": "Every week"},
          {"id": "3", "title": "Every 2 weeks"},
          {"id": "4", "title": "Once a month"}
        ]}
    ]
  },
  {
    "id": "345", "title": "Pets",
    "serviceProviders": [],
    "questions": [],
    "services": [
    {"id": "1", "title": "Cat Grooming"},
    {"id": "2", "title": "Dog Training"},
    {"id": "3", "title": "Pet Boarding"},
    {"id": "4", "title": "Pet Sitting"},
    {"id": "5", "title": "Dog Walking"},
    {
      "id": "6",
      "title": "Dog Grooming",
      "serviceProviders": [
        {
          "title": "Lucky Dogs Daycare & Pet Spa",
          "rating": 4.4,
          "reviews": 21,
          "yearsInBusiness": "7",
          "hires": 17,
          "latestReview": "This weekend we  took her for a two-hour play time followed by ...",
          "price": "$35/Dog"
        },
        {
          "title": "Zenful Pawz LLC",
          "rating": 5,
          "reviews": 1,
          "yearsInBusiness": "1",
          "hires": 3,
          "latestReview": "! Sheâ€™s the best! I know my little Betty is in great care. We have been to many groomers & this is the best yet! We will continue to visit Izzy!",
          "price": "contact for price"
        },
        {
          "title": "Fluff & Buff Boutique",
          "rating": 4.8,
          "reviews": 35,
          "yearsInBusiness": "5",
          "hires": 23,
          "latestReview": "Kate really knows how to handle dogs. Her love of animals shines and she takes great pride in her work. My dog always looks great after her care.",
          "price": "contact for price"
        }
      ],
      "questions": [
        {"id": "1", "question": "Dog size",
          "questionType": "MULTIPLE-CHOICE",
          "choices": [
            {"id": "1", "title": "Extra small (1 - 15 lbs)"},
            {"id": "2", "title": "Small (16 - 25 lbs)"},
            {"id": "3", "title": "Medium (26 - 40 lbs)"},
            {"id": "4", "title": "Large (41 - 70 lbs)"},
            {"id": "5", "title": "Extra large (70+ lbs)"}
          ]},
        {"id": "2", "question": "Dog coat type",
          "questionType": "MULTIPLE-CHOICE",
          "choices": [
            {"id": "1", "title": "Smooth, skin-like coat"},
            {"id": "2", "title": "Wavy or curly coat"},
            {"id": "3", "title": "Silky coat with fine hair"},
            {"id": "4", "title": "Wiry, bristly coat"},
            {"id": "5", "title": "Thick, fluffy double-coat"}
          ]}
      ]},
    {"id": "7", "title": "Dog Daycare"},
    {"id": "8", "title": "Aquarium Services"},
    {"id": "9", "title": "Animal Training (Non-canine)"},
    {"id": "10", "title": "Behavior Modification (Non-canine)"}
  ]},
  {
    "id": "234", "title": "Wellness",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "456", "title": "Business",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "567", "title": "Events",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "678", "title": "Crafts",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "789", "title": "Design and Web",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "890", "title": "Legal",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "987", "title": "Personal",
    "serviceProviders": [],
    "questions": [],
    "services": []},
  {
    "id": "876", "title": "Photography",
    "serviceProviders": [],
    "questions": [],
    "services": []}
]