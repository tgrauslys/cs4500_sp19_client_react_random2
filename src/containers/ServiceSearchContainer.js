import React from 'react'
import ServiceSearchBar from '../components/ServiceSearchBar';
import ServiceSearchResults from '../components/ServiceSearchResults';
import ServiceSearchFilters from '../components/ServiceSearchFilters';

class ServiceSearchContainer extends React.Component {
    constructor(props) {
        super(props)
        this.serviceService = this.props.serviceService
        this.userService = this.props.userService
        this.questionService = this.props.questionService
        if (this.props.match !== undefined) {
            this.serviceId = this.props.match.params.id
        }
        this.state = {
            serviceCategory: "",
            serviceQuestions: [],
            activeFilters: [],
            filterQuestions: [],
            searchResults: [],
            username: "",
            zipcode: ""
        }
    }
    
    componentDidMount() {
        if (this.serviceId !== undefined) {
        this.userService
            .filterUsersByService(this.serviceId, this.state.username, this.state.zipcode, [])
            .then(searchResults => {
                if (searchResults !== undefined) {
                    this.setState({
                        searchResults: searchResults
                    })
                }
            })
        this.serviceService
            .findServiceById(this.serviceId)
            .then(service => {
                let mappedQuestions = service.serviceQuestions.map(serviceQuestion => {
                        let displayedQuestion = {}
                        displayedQuestion["id"] = serviceQuestion.id
                        displayedQuestion["type"]= serviceQuestion.type
                        displayedQuestion["question"]= serviceQuestion.question
                        displayedQuestion["choices"]= serviceQuestion.choices.split(',')
                        return displayedQuestion
                    })
                this.setState({
                    serviceQuestions: service.serviceQuestions,
                    filterQuestions: mappedQuestions
                })
            })
        } else {
            this.userService
                .filterUsers(this.state.username, this.state.zipcode)
                .then(searchResults => {
                    if (searchResults !== undefined){
                        this.setState({
                            searchResults: searchResults
                        })
                    }
                })
        }
    } 
    handleSubmit = () => {
        const searchPredicates = []
        this.state.activeFilters.forEach(activeFilter => {
            const serviceQuestion = {}
            serviceQuestion["type"] = activeFilter.serviceQuestion["type"]
            serviceQuestion["question"]= activeFilter.serviceQuestion["question"]
            serviceQuestion["choices"]= activeFilter.serviceQuestion["choices"]
            const searchPredicate = {
                serviceQuestion,
                serviceAnswer: activeFilter.serviceAnswer
            }
            searchPredicates.push(searchPredicate)
        })
        this.userService
            .filterUsers(this.serviceId, this.state.username, this.state.zipcode, searchPredicates)
            .then(searchResults => {
                if (searchResults !== undefined) {
                    this.setState({
                        searchResults: searchResults
                    })
                }
            })
    }
    updateUserName = e => {
        this.setState({
            username: e.target.value
        })
    }
    updateZipcode = e => {
        this.setState({
            zipcode: e.target.value
        })
    }
    updateFilter = e => {
        let {value, name} = e.target
        let updatedFilters = this.state.activeFilters
        let questionFound = false
        for (let i = 0; i < updatedFilters.length; i++) {
            let {serviceQuestion, serviceAnswer} = updatedFilters[i]
            if (serviceQuestion.id == name) {
                questionFound = true
                switch(serviceQuestion.type) {
                    case "TrueFalse":
                        serviceAnswer["trueFalseAnswer"] = value
                        break
                    case "Range":
                        serviceAnswer["minRangeAnswer"] = value
                        serviceAnswer["maxRangeAnswer"] = value
                        break
                    case "MultipleChoice":
                        serviceAnswer["choiceAnswer"] = value
                        break
                }
            }
        }
        if (!questionFound) {
            const serviceAnswer = {}
            let serviceQuestion = null
            for (let i = 0; i < this.state.serviceQuestions.length; i++) {
                if (this.state.serviceQuestions[i].id == name) {
                    serviceQuestion = this.state.serviceQuestions[i]
                    break
                }
            }
            switch(serviceQuestion.type) {
                case "TrueFalse":
                    serviceAnswer["trueFalseAnswer"] = value
                    break
                case "Range":
                    serviceAnswer["minRangeAnswer"] = value
                    serviceAnswer["maxRangeAnswer"] = value
                    break
                case "MultipleChoice":
                    serviceAnswer["choiceAnswer"] = value
                    break
            }
            updatedFilters.push({
                serviceQuestion,
                serviceAnswer
            })

        }
            
        this.setState({
            activeFilters: updatedFilters
        }, this.handleSubmit())
    }
    render() {
        
        return (
            <div>
                <ServiceSearchBar 
                    handleSubmit={this.handleSubmit}
                    updateUsername={this.updateUserName}
                    updateZipcode={this.updateZipcode}/>
                {this.serviceId !== undefined && <ServiceSearchFilters
                    serviceSearchFilters= {this.state.filterQuestions}
                    handleSelection = {this.updateFilter}/>}
                <ServiceSearchResults
                    searchResults={this.state.searchResults}/>
            </div>
        )
    }
}

export default ServiceSearchContainer