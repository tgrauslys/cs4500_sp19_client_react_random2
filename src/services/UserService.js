export default class UserService {
    static instance = null;
    static getInstance() {
        if(UserService.instance === null) {
            UserService.instance = new UserService()
        }
        return this.instance
    }
    findUserById = userId =>
        fetch(`http://localhost:8080/api/users/${userId}`)
            .then(response => response.json())
    findAllUsers = () =>
        fetch("http://localhost:8080/api/users")
            .then(response => response.json())
}
