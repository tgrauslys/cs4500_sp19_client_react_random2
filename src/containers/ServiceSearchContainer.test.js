import React from 'react'
import ServiceSearchContainer from './ServiceSearchContainer'
import UserService from '../services/UserService'
import '../services/UserService.mock'
import { BrowserRouter } from "react-router-dom";
import TestRenderer from 'react-test-renderer'

const userService = UserService.getInstance()

test('Render provider search correctly', async () => {
    const testRenderer = TestRenderer.create(
        <BrowserRouter>
            <ServiceSearchContainer 
                id={undefined}
                service={userService}/>
        </BrowserRouter>
    )
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()

    let testInstance = testRenderer.root
    let usernameInput = testInstance.findByProps({className: "input username"})
    let zipcodeInput = testInstance.findByProps({className: "input zipcode"})
    let searchButton = testInstance.findByProps({className: "search-button btn btn-success"})

    usernameInput.props.onChange({
        target: {value: "1"}
    })

    zipcodeInput.props.onChange({
        target: {value: "02115"}
    })

    await searchButton.props.onClick({
        preventDefault: () => {}
    })
    
    tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()
})