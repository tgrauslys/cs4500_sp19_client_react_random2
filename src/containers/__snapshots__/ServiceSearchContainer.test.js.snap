// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render provider search correctly 1`] = `
<div>
  <div>
    <form
      className="form"
    >
      <div
        className="input-group mb-3"
      >
        <input
          className="input username"
          name="username"
          onChange={[Function]}
          placeholder="Search for providers..."
          type="text"
        />
        <input
          className="input zipcode"
          name="username"
          onChange={[Function]}
          placeholder="Zipcode"
          type="text"
        />
        <div
          className="input-group-append"
        >
          <button
            className="search-button btn btn-success"
            onClick={[Function]}
            style={
              Object {
                "margin": "2px",
              }
            }
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </form>
  </div>
  <div>
    <h3>
      Service Providers
    </h3>
    <table
      className="table"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Render provider search correctly 2`] = `
<div>
  <div>
    <form
      className="form"
    >
      <div
        className="input-group mb-3"
      >
        <input
          className="input username"
          name="username"
          onChange={[Function]}
          placeholder="Search for providers..."
          type="text"
        />
        <input
          className="input zipcode"
          name="username"
          onChange={[Function]}
          placeholder="Zipcode"
          type="text"
        />
        <div
          className="input-group-append"
        >
          <button
            className="search-button btn btn-success"
            onClick={[Function]}
            style={
              Object {
                "margin": "2px",
              }
            }
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </form>
  </div>
  <div>
    <h3>
      Service Providers
    </h3>
    <table
      className="table"
    >
      <tbody>
        <tr>
          <td>
            username1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
